ownCloudAppFramework/SDK Extensions/OCQueryCondition+SearchSegmenter.m

```diff
@@ -0,0 +1,304 @@
+ *
+ */
+
+#import "OCQueryCondition+SearchSegmenter.h"
+#import "NSDate+ComputedTimes.h"
+#import "NSString+ByteCountParser.h"
+
+@implementation NSString (SearchSegmenter)
+
+- (NSArray<NSString *> *)segmentedForSearch
+{
+	NSMutableArray<NSString *> *segments = [NSMutableArray new];
+	NSArray<NSString *> *terms;
+
+	if ((terms = [self componentsSeparatedByString:@" "]) != nil)
+	{
+		__block NSString *segmentString = nil;
+		__block BOOL segmentOpen = NO;
+
+		void (^SubmitSegment)(void) = ^{
+			if (segmentString.length > 0)
+			{
+				[segments addObject:segmentString];
+			}
+
+			segmentString = nil;
+		};
+
+		for (NSString *inTerm in terms)
+		{
+			NSString *term = inTerm;
+			BOOL closingSegment = NO;
+
+			if ([term hasPrefix:@"\""])
+			{
+				// Submit any open segment
+				SubmitSegment();
+
+				// Start new segment
+				term = [term substringFromIndex:1];
+				segmentOpen = YES;
+			}
+
+			if ([term hasSuffix:@"\""])
+			{
+				// End segment
+				term = [term substringToIndex:term.length-1];
+				closingSegment = YES;
+			}
+
+			// Append term to current segment
+			if (segmentString.length == 0)
+			{
+				segmentString = term;
+
+				if (!segmentOpen)
+				{
+					// Submit standalone segment
+					SubmitSegment();
+				}
+			}
+			else
+			{
+				// Append to segment string
+				segmentString = [segmentString stringByAppendingFormat:@" %@", term];
+			}
+
+			// Submit closed segments
+			if (closingSegment)
+			{
+				segmentOpen = NO;
+				SubmitSegment();
+			}
+		}
+
+		SubmitSegment();
+	}
+
+	return (segments);
+}
+
+@end
+
+@implementation OCQueryCondition (SearchSegmenter)
+
++ (instancetype)forSearchSegment:(NSString *)segmentString
+{
+	NSString *segmentStringLowercase = segmentString.lowercaseString;
+
+	if ([segmentStringLowercase isEqual:@":folder"])
+	{
+		return ([OCQueryCondition where:OCItemPropertyNameType isEqualTo:@(OCItemTypeCollection)]);
+	}
+	else if ([segmentStringLowercase isEqual:@":file"])
+	{
+		return ([OCQueryCondition where:OCItemPropertyNameType isEqualTo:@(OCItemTypeFile)]);
+	}
+	else if ([segmentStringLowercase isEqual:@":image"])
+	{
+		return ([OCQueryCondition where:OCItemPropertyNameMIMEType startsWith:@"image/"]);
+	}
+	else if ([segmentStringLowercase isEqual:@":video"])
+	{
+		return ([OCQueryCondition where:OCItemPropertyNameMIMEType startsWith:@"video/"]);
+	}
+	else if ([segmentStringLowercase isEqual:@":today"])
+	{
+		return ([OCQueryCondition where:OCItemPropertyNameLastModified isGreaterThan:[NSDate startOfRelativeDay:0]]);
+	}
+	else if ([segmentStringLowercase isEqual:@":week"])
+	{
+		return ([OCQueryCondition where:OCItemPropertyNameLastModified isGreaterThan:[NSDate startOfRelativeWeek:0]]);
+	}
+	else if ([segmentStringLowercase isEqual:@":month"])
+	{
+		return ([OCQueryCondition where:OCItemPropertyNameLastModified isGreaterThan:[NSDate startOfRelativeMonth:0]]);
+	}
+	else if ([segmentStringLowercase isEqual:@":year"])
+	{
```

Comment:
The keywords below supports localized strings. Shouldn't this available for the strings `folder`, `file`, `image`, `video`, `today`, `week`, `month`, `year` too?

Reply:
I missed those. The latest commit allows localizing these, too, but now uses a normalization + localization step to both simplify the code and prepare the next steps where queries could always be saved in English to allow a common syntax across client and server.